src/protocol/reveal_additive_binary.rs

```diff
@@ -0,0 +1,102 @@
+use crate::{
+    error::BoxError,
+    helpers::{fabric::Network, prss::SpaceIndex, Direction},
+    protocol::{context::ProtocolContext, RecordId, Step},
+};
+
+use futures::future::try_join;
+use serde::{Deserialize, Serialize};
+
+/// A message sent by each helper when they've revealed their own shares
+#[allow(clippy::module_name_repetitions)]
+#[derive(Debug, Serialize, Deserialize, Default)]
+pub struct RevealValue {
+    share: bool,
+}
+
+/// This implements reveal algorithm
+/// As this is an additive sharing, each helper has just one boolean share
+/// As such, reveal requires each helper to send their share both left and right
+/// Put another way, each `P_i` sends `\[a\]_i` to `P_i+1` and `P_i-1`
+/// and then reconstructs a from` \[a\]_i`, `\[a\]_i+i` and `\[a\]i−1`.
+// Input: Each helpers know their own secret shares
+// Output: At the end of the protocol, all 3 helpers know a revealed (or opened) secret
+#[derive(Debug)]
+pub struct RevealAdditiveBinary {}
+
+impl RevealAdditiveBinary {
+    #[allow(dead_code)]
+    pub async fn execute<S: Step + SpaceIndex, N: Network<S>>(
+        ctx: &ProtocolContext<'_, S, N>,
+        step: S,
+        record_id: RecordId,
+        input: bool,
+    ) -> Result<bool, BoxError> {
+        let mut channel = ctx.gateway.get_channel(step);
+
+        let send_left_future = channel.send(
+            channel.identity().peer(Direction::Left),
+            record_id,
+            RevealValue { share: input },
+        );
+
+        let send_right_future = channel.send(
+            channel.identity().peer(Direction::Right),
+            record_id,
+            RevealValue { share: input },
+        );
+
+        try_join(send_left_future, send_right_future).await?;
```

Comment:
This *does not work*. 

The error is:

```
error[E0499]: cannot borrow `channel` as mutable more than once at a time
  --> src/protocol/reveal_additive_binary.rs:43:33
   |
37 |           let send_left_future = channel.send(
   |  ________________________________-
38 | |             channel.identity().peer(Direction::Left),
39 | |             record_id,
40 | |             RevealValue { share: input },
41 | |         );
   | |_________- first mutable borrow occurs here
42 |
43 |           let send_right_future = channel.send(
   |  _________________________________^
44 | |             channel.identity().peer(Direction::Right),
45 | |             record_id,
46 | |             RevealValue { share: input },
47 | |         );
   | |_________^ second mutable borrow occurs here
48 |
49 |           try_join(send_left_future, send_right_future).await?;
   |                    ---------------- first borrow later used here
```

Is this something I can work around? Or does this require an infra-layer change?

Reply:
oh, that's my bad - infra does not support it yet. #132 should fix it